src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+
+		for (i = 0; i < 2; i++)
+		{
+			if (sp->std[i].fd != -1)
+			{
+			#ifdef G_OS_WIN32
+				sp->ioc[i] = g_io_channel_win32_new_fd(sp->std[i].fd);
+			#else
+				sp->ioc[i] = g_io_channel_unix_new(sp->std[i].fd);
+				g_source_add_poll(source, &sp->std[i]);
+			#endif
+				g_io_channel_set_encoding(sp->ioc[i], NULL, NULL);
+			}
+		}
+
+		sp->exit_cb = exit_cb;
+		sp->exit_data = exit_data;
+
+		if (stdin_text || stdout_cb || stderr_cb)
+		{
+			g_source_attach(source, NULL);
+			g_source_unref(source);
+		}
+		else
+			g_child_watch_add(sp->pid, spawn_source_exit_cb, sp);
+
+		if (block_window)
+		{
+			sp->dialog = gtk_message_dialog_new(block_window, GTK_DIALOG_DESTROY_WITH_PARENT,
+				GTK_MESSAGE_OTHER, GTK_BUTTONS_CANCEL, _("Please wait"));
+			sp->command = command_line ? command_line : argv[0];
+			gtk_message_dialog_format_secondary_text(GTK_MESSAGE_DIALOG(sp->dialog),
+				_("Running %s..."), sp->command);
+			g_signal_connect(sp->dialog, "response", G_CALLBACK(spawn_dialog_response), sp);
+			gtk_dialog_run(GTK_DIALOG(sp->dialog));
+		}
+
+		return TRUE;
+	}
+
+	g_source_unref(source);
+	return FALSE;
+}
+
+
+static void spawn_capture_text(const gchar *text, gpointer user_data)
+{
+	g_string_append((GString *) user_data, text);
+}
+
+
+typedef struct _SpawnCapture
+{
+	gchar **stdout_text;
+	GString *stdout_str;
+	gchar **stderr_text;
+	GString *stderr_str;
+	gint *exit_status;
+} SpawnCapture;
+
+
+static void spawn_capture_exit(gint status, gpointer user_data)
+{
+	SpawnCapture *sp = (SpawnCapture *) user_data;
+
+	if (sp->stdout_text)
+		*sp->stdout_text = g_string_free(sp->stdout_str, FALSE);
+
+	if (sp->stderr_text)
+		*sp->stderr_text = g_string_free(sp->stderr_str, FALSE);
+
+	if (sp->exit_status)
+		*sp->exit_status = status;
+
+	g_free(user_data);
+}
+
+
+/**
+ *  Executes a child program and captures it's output.
+ *
+ *  Either a command line or an argument vector must be passed. If both are present, the
+ *  argument vector is appended to the command line. An empty command line is not allowed.
+ *
+ *  If a @a block_window is specified (recommended), the function will not return until the
+ *  child exits, but all other windows (if any) will receive events. Otherwise it'll return,
+ *  and it's your responsibility to keep @a stdout_text, @a stderr_text and @a exit_status
+ *  valid until the execution finishes.
+ *
+ *  @a stdin_text will be gfree()-d automatically when it's sent entirely to the child, or
+ *  when the child exits.
+ *
+ *  @a stdout_text and @a stderr_text will be set when the child exits.
+ *
+ *  @param working_directory child's current working directory, or @c NULL.
+ *  @param command_line child program and arguments, or @c NULL.
+ *  @param argv child's argument vector, or @c NULL.
+ *  @param envp child's environment, or @c NULL.
+ *  @param block_window window whose message loop to block until the child exits, or @c NULL.
+ *  @param stdin_text text to send to childs's stdin, or @c NULL.

```

Comment:
No I think stealing it is fine, it's unlikely callers didn't allocate it or need to keep it around -- and if they do they can always g_strdup() it themselves.  I just think this should be mentioned in the argument description too to make it a little more obvious just in case (although the fact the argument is non-constant already suggests it's either modified or stolen).

Ideally I guess the function would add a reference to the string and the caller would drop it if he wants to, but well, this is C so what are you gonna do.